service/src/Helper.cpp

```diff
@@ -268,4 +270,148 @@ namespace geopm
         }
         return stat_struct.st_gid;
     };
+
+    int pidfd_open(pid_t pid, int flags)
+    {
+        return syscall(SYS_pidfd_open, pid, flags);
+    }
+
+    pid_t pidfd_to_pid(int pidfd)
```

Comment:
When do we plan on translating pidfd to its self-viewed pid?

If we plan to translate pidfds to pids for use in our existing control mechanisms, then I think we'd need to include the namespaces themselves as part of the identifier. Or maybe just use the pidfd? Otherwise, we need to get the process's PID from a common ancestor namespace instead of getting it from the target process's namespace (to avoid collisions between same PID numbers in different namespaces).

Reply:
You can change PID namespaces with the pidfd using the setns(2) call.  This is the intended use of the pidfd_ns paramters.

Reply:
An example of my concern is if two separate applications attach to the GEOPM service from separate containers. If app1's SID is 1 in its PID namespace and app2's SID is also 1 in its PID namespace, then GEOPM would wrongly think they are in the same session (if we only use self-viewed PIDs as identifiers), even if they are actually different processes.

Reply:
We will store the pidfd for the session leader PID.  This pidfd will always be used with setns() prior to calling getsid() on the user provided pid.  We wouldn't use the pidfd namespace of the user provided pid (the problematic case you are referring to).  The session leader PID namespace would always be the reference namespace when determining another pid's session leader.